File: tests/cli/cmd_etl/batetl/test___main__.py
#!/usr/bin/env python3
#
# Copyright (c) 2010 - 2025, Fraunhofer-Gesellschaft zur Foerderung der angewandten Forschung e.V.
# All rights reserved.
#
# SPDX-License-Identifier: BSD-3-Clause
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are met:
#
# 1. Redistributions of source code must retain the above copyright notice, this
#    list of conditions and the following disclaimer.
#
# 2. Redistributions in binary form must reproduce the above copyright notice,
#    this list of conditions and the following disclaimer in the documentation
#    and/or other materials provided with the distribution.
#
# 3. Neither the name of the copyright holder nor the names of its
#    contributors may be used to endorse or promote products derived from
#    this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
# FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# We kindly request you to use one or more of the following phrases to refer to
# foxBMS in your hardware, software, documentation or advertising materials:
#
# - "This product uses parts of foxBMS®"
# - "This product includes parts of foxBMS®"
# - "This product is derived from foxBMS®"

"""Testing file 'cli/cmd_etl/batetl/__main__.py'."""

import io
import sys
import unittest
from contextlib import redirect_stdout
from pathlib import Path
from unittest.mock import MagicMock, Mock, call, patch

import yaml
from click.testing import CliRunner

try:
    from cli.cmd_etl.batetl import (
        DEFAULT_CONFIG_FILE_FILTER,
        cmd_decode,
        cmd_filter,
        cmd_table,
        get_config_file_options,
        main,
    )
    from cli.cmd_etl.batetl.__main__ import entry_point
except ModuleNotFoundError:
    sys.path.insert(0, str(Path(__file__).parents[4]))
    from cli.cmd_etl.batetl import (
        DEFAULT_CONFIG_FILE_FILTER,
        cmd_decode,
        cmd_filter,
        cmd_table,
        get_config_file_options,
        main,
    )
    from cli.cmd_etl.batetl.__main__ import entry_point


class TestBatetlMain(unittest.TestCase):
    """Test batetl package entry point"""

    def test_entry_point(self) -> None:
        """check help message output of the main command"""
        runner = CliRunner()
        result = runner.invoke(entry_point, [])
        self.assertTrue("Usage" in result.output)


@patch("cli.cmd_etl.batetl.read_config")
class TestBatetlGetConfigFileOptions(unittest.TestCase):
    """Test batetl configuration parser"""

    def test_get_config_file_options_no_file(self, _: MagicMock) -> None:
        """Test when the configuration file does not exist."""
        ctx = MagicMock()
        config_file = MagicMock()
        config_file.is_file.return_value = False
        get_config_file_options(ctx, MagicMock(), config_file)
        self.assertDictEqual(ctx.default_map, {})

    def test_get_config_file_options_file_exists_no_sampling(
        self, m_read_config: MagicMock
    ) -> None:
        """Test when the configuration file exists."""
        ctx = MagicMock()
        config_file = MagicMock()
        config_file.is_file.return_value = True
        m_read_config.return_value = {"foo": "bar"}

        get_config_file_options(ctx, MagicMock(), config_file)
        self.assertDictEqual(ctx.default_map, {"foo": "bar"})

    def test_get_config_file_options_file_exists_sampling(
        self, m_read_config: MagicMock
    ) -> None:
        """Test when the configuration file exists and sampling is provided."""
        ctx = MagicMock()
        config_file = MagicMock()
        config_file.is_file.return_value = True
        m_read_config.return_value = {"foo": "bar", "sampling": {"a": "b"}}

        get_config_file_options(ctx, MagicMock(), config_file)
        self.assertDictEqual(
            ctx.default_map, {"foo": "bar", "sampling": {"a": "b"}.items()}
        )

    @patch("cli.cmd_etl.batetl.Path.is_file")
    def test_get_config_file_options_default_file(
        self, m_is_file: MagicMock, m_read_config: MagicMock
    ) -> None:
        """Test when the configuration file exists."""
        ctx = MagicMock()
        config_file = DEFAULT_CONFIG_FILE_FILTER
        m_is_file.return_value = True
        m_read_config.return_value = {"foo": "bar"}
        buf = io.StringIO()
        with redirect_stdout(buf):
            get_config_file_options(ctx, MagicMock(), config_file)
        self.assertIn("Default configuration file filter.yml is used.", buf.getvalue())


class TestBatetlDecode(unittest.TestCase):
    """Test 'decode' command"""

    def test_cmd_decode_help(self) -> None:
        "Test help option"
        runner = CliRunner()
        result = runner.invoke(cmd_decode, [])
        self.assertTrue("Usage" in result.output)

    @patch("cli.cmd_etl.batetl.run_decode2file")
    @patch("cli.cmd_etl.batetl.run_decode2stdout")
    @patch("cli.cmd_etl.batetl.can_decode_setup")
    def test_cmd_decode_to_stdout(
        self,
        m_can_decode_setup: Mock,
        run_decode2stdout_mock: Mock,
        run_decode2file_mock: Mock,
    ) -> None:
        """Tests the decode command line interface

        :param setup_mock: A mock for the CANDecode class setup
        :param run_decode_mock: A Mock for the run_decode function, which
            executes the decoding
        """
        # Case 1: check execution with stdout as output stream
        decode_obj_mock = Mock(name="decode")
        m_can_decode_setup.return_value = decode_obj_mock
        runner = CliRunner()
        with runner.isolated_filesystem():
            p = Path("test.yml")

            p.write_text(
                yaml.dump(
                    {
                        "dbc": str(Path(__file__)),
                        "timestamp_pos": 0,
                        "id_pos": 1,
                        "data_pos": 2,
                    }
                ),
                encoding="utf-8",
            )
            result = runner.invoke(cmd_decode, ["--config", str(p.resolve())])
            m_can_decode_setup.assert_called_once_with(
                {
                    "dbc": Path(__file__),
                    "timestamp_pos": 0,
                    "id_pos": 1,
                    "data_pos": 2,
                    "output": None,
                }
            )
            run_decode2stdout_mock.assert_called_once_with(decode_obj_mock)
            run_decode2file_mock.assert_not_called()
            self.assertEqual(result.exit_code, 0)
        m_can_decode_setup.reset_mock()
        run_decode2stdout_mock.reset_mock()
        run_decode2file_mock.reset_mock()

    @patch("cli.cmd_etl.batetl.run_decode2file")
    @patch("cli.cmd_etl.batetl.run_decode2stdout")
    @patch("cli.cmd_etl.batetl.can_decode_setup")
    def test_cmd_decode_to_files(
        self,
        m_can_decode_setup: Mock,
        run_decode2stdout_mock: Mock,
        run_decode2file_mock: Mock,
    ) -> None:
        """Tests the decode command line interface

        :param setup_mock: A mock for the CANDecode class setup
        :param run_decode_mock: A Mock for the run_decode function, which
            executes the decoding
        """
        # Case 1: check execution with stdout as output stream
        decode_obj_mock = Mock(name="decode")
        m_can_decode_setup.return_value = decode_obj_mock
        runner = CliRunner()
        with runner.isolated_filesystem():
            p = Path("test.yml")

            p.write_text(
                yaml.dump(
                    {
                        "dbc": str(Path(__file__)),
                        "timestamp_pos": 0,
                        "id_pos": 1,
                        "data_pos": 2,
                        "output": str(Path(__file__).parent),
                    }
                ),
                encoding="utf-8",
            )
            result = runner.invoke(cmd_decode, ["--config", str(p.resolve())])
            m_can_decode_setup.assert_called_once_with(
                {
                    "dbc": Path(__file__),
                    "timestamp_pos": 0,
                    "id_pos": 1,
                    "data_pos": 2,
                    "output": Path(__file__).parent,
                }
            )
            run_decode2stdout_mock.assert_not_called()
            run_decode2file_mock.assert_called_once_with(decode_obj_mock)
            self.assertEqual(result.exit_code, 0)
        m_can_decode_setup.reset_mock()
        run_decode2stdout_mock.reset_mock()
        run_decode2file_mock.reset_mock()


class TestBatetlFilter(unittest.TestCase):
    """Test 'filter' command"""

    def test_cmd_filter_help(self) -> None:
        "Test help option"
        runner = CliRunner()
        result = runner.invoke(cmd_filter, ["-c", "test"])
        self.assertTrue("Usage" in result.output)

    @patch("cli.cmd_etl.batetl.run_filter")
    @patch("cli.cmd_etl.batetl.can_filter_setup")
    def test_cmd_filter(self, setup_mock: Mock, run_filter_mock: Mock) -> None:
        """Tests the filter command line interface

        :param setup_mock: A mock for the CANFilter class setup
        :param run_filter_mock: A Mock for the run_filter function, which
            executes the filtering
        """
        filter_obj_mock = Mock(name="filter")
        setup_mock.return_value = filter_obj_mock
        runner = CliRunner()
        with runner.isolated_filesystem():
            with open("filter.yml", mode="w", encoding="utf-8") as f:
                f.write('id_pos: 4\nids: ["0241", "035C"]')
            result = runner.invoke(cmd_filter, ["--config", "filter.yml"])
            setup_mock.assert_called_once_with(
                {
                    "id_pos": 4,
                    "ids": ["0241", "035C"],
                    "sampling": {},
                    "_input": None,
                    "output": None,
                }
            )
            run_filter_mock.assert_called_once_with(filter_obj_mock)
            self.assertEqual(result.exit_code, 0)

    @patch("cli.cmd_etl.batetl.run_filter")
    @patch("cli.cmd_etl.batetl.can_filter_setup")
    def test_cmd_filter_using_sampling(
        self, setup_mock: Mock, run_filter_mock: Mock
    ) -> None:
        """Tests the filter command line interface

        :param setup_mock: A mock for the CANFilter class setup
        :param run_filter_mock: A Mock for the run_filter function, which
            executes the filtering
        """
        filter_obj_mock = Mock(name="filter")
        setup_mock.return_value = filter_obj_mock
        runner = CliRunner()
        with runner.isolated_filesystem():
            with open("filter.yml", mode="w", encoding="utf-8") as f:
                f.write('id_pos: 4\nids: ["0241", "035C"]\n"sampling":\n  "240": 10')
            result = runner.invoke(cmd_filter, ["--config", "filter.yml"])
            setup_mock.assert_called_once_with(
                {
                    "id_pos": 4,
                    "ids": ["0241", "035C"],
                    "sampling": {"240": 10},
                    "_input": None,
                    "output": None,
                }
            )
            run_filter_mock.assert_called_once_with(filter_obj_mock)
            self.assertEqual(result.exit_code, 0)


class TestBatetlTable(unittest.TestCase):
    """Test 'table' command"""

    def test_cmd_table_help(self) -> None:
        "Test help option for the 'table' command when no argument or option are provided"
        runner = CliRunner()
        result = runner.invoke(cmd_table, [])
        self.assertTrue("Usage" in result.output)

    @patch("cli.cmd_etl.batetl.run_table")
    @patch("cli.cmd_etl.batetl.table_setup")
    def test_cmd_table(self, setup_mock: Mock, run_table_mock: Mock) -> None:
        """Tests the table command line interface

        :param setup_mock: A mock for the Table class setup
        :param run_table_mock: A Mock for the run_table function, which
            executes the create of a table out of the CAN data
        """
        # Case 1: check execution as expected
        table_obj_mock = Mock(name="table")
        setup_mock.return_value = table_obj_mock
        runner = CliRunner()
        with runner.isolated_filesystem():
            with open("test_config.txt", mode="w", encoding="utf-8") as _:
                with open("test_output.txt", mode="w", encoding="utf-8") as _:
                    with open("test_data.txt", mode="w", encoding="utf-8") as _:
                        result = runner.invoke(
                            cmd_table,
                            [
                                "--config",
                                "test_config.txt",
                                "--output",
                                "test_output.txt",
                                "test_data.txt",
                            ],
                        )
                        setup_mock.assert_called_once_with(Path("test_config.txt"))
                        run_table_mock.assert_called_once_with(
                            table_obj_mock,
                            Path("test_data.txt"),
                            Path("test_output.txt"),
                        )
                        self.assertEqual(result.exit_code, 0)


class TestBatetlMisc(unittest.TestCase):
    """Test the wrapper for the cli integration"""

    @patch("cli.cmd_etl.batetl.entry_point")
    def test_main(self, m_entry_point: MagicMock) -> None:
        """Tests the main function in command line interface

        :param setup_mock: A mock for the entry point function
        """
        m_entry_point.add_command = MagicMock()
        main()
        m_entry_point.add_command.assert_has_calls(
            [call(cmd_filter), call(cmd_decode), call(cmd_table)]
        )


if __name__ == "__main__":
    unittest.main()

File: tests/unit/bootloader/main/test_main.c
/**
 *
 * @copyright &copy; 2010 - 2025, Fraunhofer-Gesellschaft zur Foerderung der angewandten Forschung e.V.
 * All rights reserved.
 *
 * SPDX-License-Identifier: BSD-3-Clause
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are met:
 *
 * 1. Redistributions of source code must retain the above copyright notice, this
 *    list of conditions and the following disclaimer.
 *
 * 2. Redistributions in binary form must reproduce the above copyright notice,
 *    this list of conditions and the following disclaimer in the documentation
 *    and/or other materials provided with the distribution.
 *
 * 3. Neither the name of the copyright holder nor the names of its
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * We kindly request you to use one or more of the following phrases to refer to
 * foxBMS in your hardware, software, documentation or advertising materials:
 *
 * - "This product uses parts of foxBMS&reg;"
 * - "This product includes parts of foxBMS&reg;"
 * - "This product is derived from foxBMS&reg;"
 *
 */

/**
 * @file    test_main.c
 * @author  foxBMS Team
 * @date    2020-04-01 (date of creation)
 * @updated 2025-08-07 (date of last update)
 * @version v1.10.0
 * @ingroup UNIT_TEST_IMPLEMENTATION
 * @prefix  TEST
 *
 * @brief   Tests of the basic initialization of the hardware done in main
 * @details This is a test of the basic initialization. However, this basic
 *          initialization is performed **after** the low-level initialization
 *          in the startup implementation.
 */

/*========== Includes =======================================================*/
#include "unity.h"
#include "MockHL_gio.h"
#include "MockHL_pinmux.h"
#include "MockHL_rti.h"
#include "MockHL_sys_core.h"
#include "Mockboot.h"
#include "Mockcan.h"
#include "Mockinfinite-loop-helper.h"
#include "Mockrti.h"

#include "boot_cfg.h"

#include "fstd_types.h"
#include "main.h"

/*========== Unit Testing Framework Directives ==============================*/
TEST_SOURCE_FILE("main.c")

TEST_INCLUDE_PATH("../../src/bootloader/driver/can")
TEST_INCLUDE_PATH("../../src/bootloader/driver/config")
TEST_INCLUDE_PATH("../../src/bootloader/driver/rti")
TEST_INCLUDE_PATH("../../src/bootloader/engine/boot")
TEST_INCLUDE_PATH("../../src/bootloader/main")

/*========== Definitions and Implementations for Unit Test ==================*/
/* Define the following variables in the following context to prevent unittest
error. */
uint32_t main_textLoadStartFlashC     = 0u;
uint32_t main_textSizeFlashC          = 0u;
uint32_t main_textRunStartFlashC      = 0u;
uint32_t main_constLoadStartFlashCfgC = 0u;
uint32_t main_constSizeFlashCfgC      = 0u;
uint32_t main_constRunStartFlashCfgC  = 0u;

/*========== Setup and Teardown =============================================*/
void setUp(void) {
}

void tearDown(void) {
}

/*========== Test Cases =====================================================*/

void testMain(void) {
    /* ======= Routine tests =============================================== */
    /* ======= RT1/11: wait if no any can request comes */
    boot_state = BOOT_FSM_STATE_WAIT;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();
    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    for (uint8_t i_call = 0u; i_call < 10u; i_call++) {
        FOREVER_ExpectAndReturn(1);
        _enable_IRQ_interrupt__Expect();
        RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
        BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_WAIT);
    }
    FOREVER_ExpectAndReturn(0);

    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_WAIT, boot_state);

    /* ======= RT2/11: normal loading process, everything runs well */
    resetTest();
    boot_state = BOOT_FSM_STATE_WAIT;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();

    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_LOAD);
    for (uint8_t i_call = 0u; i_call < 8u; i_call++) {
        FOREVER_ExpectAndReturn(1);
        _enable_IRQ_interrupt__Expect();
        RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
        BOOT_GetBootStateDuringLoad_ExpectAndReturn(BOOT_FSM_STATE_LOAD);
    }
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootStateDuringLoad_ExpectAndReturn(BOOT_FSM_STATE_WAIT);
    FOREVER_ExpectAndReturn(0);

    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_WAIT, boot_state);

    /* ======= RT3/11: normal loading process, error appears during loading */
    resetTest();
    boot_state = BOOT_FSM_STATE_WAIT;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();

    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_LOAD);
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootStateDuringLoad_ExpectAndReturn(BOOT_FSM_STATE_ERROR);
    for (uint8_t i_call = 0u; i_call < 8u; i_call++) {
        FOREVER_ExpectAndReturn(1);
        _enable_IRQ_interrupt__Expect();
        RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
        BOOT_GetBootStateDuringError_ExpectAndReturn(BOOT_FSM_STATE_ERROR);
    }
    FOREVER_ExpectAndReturn(0);

    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_ERROR, boot_state);

    /* ======= RT4/11: normal loading process, error appears during loading,
    incoming reset can request to correct the error, and the reset operation is
    successful */
    resetTest();
    boot_state = BOOT_FSM_STATE_WAIT;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();

    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_LOAD);
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootStateDuringLoad_ExpectAndReturn(BOOT_FSM_STATE_ERROR);
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootStateDuringError_ExpectAndReturn(BOOT_FSM_STATE_RESET);
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_ResetBootloader_ExpectAndReturn(STD_OK);
    for (uint8_t i_call = 0u; i_call < 6u; i_call++) {
        FOREVER_ExpectAndReturn(1);
        _enable_IRQ_interrupt__Expect();
        RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
        BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_WAIT);
    }
    FOREVER_ExpectAndReturn(0);

    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_WAIT, boot_state);

    /* ======= RT5/11: normal loading process, error appears during loading,
    incoming reset request to correct the error, but the reset operation is
    not successful */
    resetTest();
    boot_state = BOOT_FSM_STATE_WAIT;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();

    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_LOAD);
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootStateDuringLoad_ExpectAndReturn(BOOT_FSM_STATE_ERROR);
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootStateDuringError_ExpectAndReturn(BOOT_FSM_STATE_RESET);
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_ResetBootloader_ExpectAndReturn(STD_NOT_OK);
    for (uint8_t i_call = 0u; i_call < 6u; i_call++) {
        FOREVER_ExpectAndReturn(1);
        _enable_IRQ_interrupt__Expect();
        RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
        BOOT_GetBootStateDuringError_ExpectAndReturn(BOOT_FSM_STATE_ERROR);
    }
    FOREVER_ExpectAndReturn(0);

    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_ERROR, boot_state);

    /* ======= RT6/11: run application, there is an valid program available and
    everything went well while jumping into the last flashed program */
    resetTest();
    boot_state = BOOT_FSM_STATE_WAIT;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();

    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_RUN);
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_IsProgramAvailableAndValidated_ExpectAndReturn(true);
    BOOT_JumpInToLastFlashedProgram_ExpectAndReturn(STD_OK);
    for (uint8_t i_call = 0u; i_call < 8u; i_call++) {
        FOREVER_ExpectAndReturn(1);
        _enable_IRQ_interrupt__Expect();
        RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
        BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_WAIT);
    }
    FOREVER_ExpectAndReturn(0);

    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_WAIT, boot_state);

    /* ======= RT7/11: run application, there is an valid program available but
    it fails while jumping into the last flashed program */
    resetTest();
    boot_state = BOOT_FSM_STATE_WAIT;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();

    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_RUN);
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_IsProgramAvailableAndValidated_ExpectAndReturn(true);
    BOOT_JumpInToLastFlashedProgram_ExpectAndReturn(STD_NOT_OK);
    for (uint8_t i_call = 0u; i_call < 8u; i_call++) {
        FOREVER_ExpectAndReturn(1);
        _enable_IRQ_interrupt__Expect();
        RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
        BOOT_GetBootStateDuringError_ExpectAndReturn(BOOT_FSM_STATE_ERROR);
    }
    FOREVER_ExpectAndReturn(0);

    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_ERROR, boot_state);

    /* ======= RT8/11: run application, there is no valid program available */
    resetTest();
    boot_state = BOOT_FSM_STATE_WAIT;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();

    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_RUN);
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    BOOT_IsProgramAvailableAndValidated_ExpectAndReturn(false);
    for (uint8_t i_call = 0u; i_call < 8u; i_call++) {
        FOREVER_ExpectAndReturn(1);
        _enable_IRQ_interrupt__Expect();
        RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
        BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_WAIT);
    }
    FOREVER_ExpectAndReturn(0);

    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_WAIT, boot_state);

    /* ======= RT9/11: unregistered state */
    resetTest();
    boot_state = 10u;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();

    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, false);
    FOREVER_ExpectAndReturn(0);

    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_ERROR, boot_state);

    /* ======= RT9/11: timeout, boot_state is BOOT_FSM_STATE_WAIT */
    resetTest();
    boot_state = BOOT_FSM_STATE_WAIT;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();

    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, true);
    rtiStopCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    BOOT_IsProgramAvailableAndValidated_ExpectAndReturn(false);
    FOREVER_ExpectAndReturn(0);

    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_WAIT, boot_state);

    /* ======= RT10/11: timeout, boot_state is BOOT_FSM_STATE_WAIT, try to run app,
    but no validated app available */
    resetTest();
    boot_state = BOOT_FSM_STATE_WAIT;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();

    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, true);
    rtiStopCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    BOOT_IsProgramAvailableAndValidated_ExpectAndReturn(false);
    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    BOOT_GetBootState_ExpectAndReturn(BOOT_FSM_STATE_WAIT);
    FOREVER_ExpectAndReturn(0);

    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_WAIT, boot_state);

    /* ======= RT11/11: timeout, boot_state is not BOOT_FSM_STATE_WAIT */
    resetTest();
    boot_state = BOOT_FSM_STATE_ERROR;

    _enable_interrupt__Expect();

    muxInit_Expect();
    gioInit_Expect();
    CAN_Initialize_Expect();
    rtiInit_Expect();

    CAN_SendBootMessage_Expect();
    RTI_ResetFreeRunningCount_ExpectAndReturn(1);
    rtiStartCounter_Expect(rtiREG1, rtiCOUNTER_BLOCK0);
    RTI_GetFreeRunningCount_ExpectAndReturn(0u);

    FOREVER_ExpectAndReturn(1);
    _enable_IRQ_interrupt__Expect();
    RTI_IsTimeElapsed_ExpectAndReturn(0u, MAIN_TIME_OUT_us, true);
    BOOT_GetBootStateDuringError_ExpectAndReturn(BOOT_FSM_STATE_ERROR);
    FOREVER_ExpectAndReturn(0);
    unit_test_main();
    TEST_ASSERT_EQUAL(BOOT_FSM_STATE_ERROR, boot_state);
}
